Add optional filter for version to is_sle

kravciak opened the pull request:
Version checking mechanism for SLE requires the use of <b>is_sle</b> with <b>sle_version_at_least</b> because sle_version_at_least returns true for all non-sle products. I think it's confusing default leading to unreadable code.

I don't want to modify / fix sle_version_at_least at once, because it's returning true by default and it might break many tests. Instead I added optional version parameter to is_sle, allowing filtering this way:
```perl
Format: [= > < >= <=] version [+]
Example:
 is_sle '=12';
 is_sle '=12-sp3';
 is_sle '>=15';
 is_sle '15+';
 is_sle '<15'
```

Our code could be then rewritten into more readable way:
```perl
Now:
if (is_sle && sle_version_at_least '15')
if (is_sle && sle_version_at_least('12-SP2') && !sle_version_at_least('15'))

After:
if (is_sle '15+')
if (is_sle('>=12-sp2') && is_sle('<15'))
```

Local runs:
 - http://dhcp91.suse.cz/tests/1443 - SLE
 - http://dhcp91.suse.cz/tests/1440 - Kubic

kravciak commented:
@rwx788 @okurz @nicksinger @Soulofdestiny

asmorodskyi commented:
what you suggestion regarding such code ? 

`is_sle && !sle_version_at_least('15') `

kravciak commented:
what about is_sle('15-')? ;)

asmorodskyi commented:
Sounds like a plan :) I also thought about something similar actually . Would you mind to implement this "-" in your PR ? 
Generally speaking I really like your idea !

kravciak commented:
@asmorodskyi - I don't mind implementing "-"

@rwx788 - what do you mean by "it doesn't belong to is_sle"? You are right, it's sle_version_at_least returning false for non-sle.

Leap would need different approach (or I can just hack and say - if it's 42.X, treat it as 14.X)
is_leap is also specific because check is based only on version number format "XX.X"
```
  return 1 if get_var('VERSION', '') =~ /[0-9]{2,}\.[0-9]/;
```
I don't want to replace it all at once, leap would be next step.

rwx788 commented:
@kravciak I mean, name is misleading, as it checks only DISTRI variable. I guess what we could do is to call sle_version_at_least deprecated, start using new function whenever possible and refactor old code not to use it. Something like check_sle_version. One of the current scenarios is that we check if it's 15 (we don't care if sle of leap), so I guess we can come up with multiple wrappers which use main logic of versions comparisons and even reuse current version_at_least methods.

The main problem here is that we cannot get rid of sle_version_at_least logic and simply introduce is_sle check there, as I see that as main issue, other things are more or less syntactic sugar. As a check solution I was thinking about wrapper which does check is_sle && sle_version_at_least not to have it everywhere.

So let's maybe try to replace all occurrences of is_sle && sle_version_at_least('15') with new method, as well as is_sle && !sle_version_at_least('15') and see how it goes. WDYT?

nicksinger commented:
Like the idea in general because it bundles two commonly used functions into one single call. But one question remains. You've provided the following parameter example: `is_sle 's390x'`
How do we want to combine such a architecture specific call with a check for a version. So let's say s390x and sle > 15? How would you implement this? `is_sle 's390x' '15+'? Or just combine it (`is_sle('s390x') '' && is_sle('15+')`)?

kravciak commented:
@nicksinger my proposal was
```
is_sle('s390x') && is_sle('15+')
```
but @rwx788 pointed me to possible solution when using hash as parameter, which would allow
```
is_sle(version => '15+', arch => 's390x')
```

As version is most used parameter I will start with is_sle('15') - accepting only version for now, hash in the future.

kravciak commented:
@asmorodskyi - I implemented "-" and I don't like it because it's suggesting "less then version 15 (INCLUDING)" - which is usually not what we want (I even made mistake first time I used it)

I would prefer to use last variant:
```
is_sle('15-')
is_sle('-15')
!is_sle('15+')
```

okurz commented:
I like it in general. I only see that it's not obvious to everyone that something like '12+' is supported and people would tend to just write '12' to *get it to work* so maybe we can make that more explicit by not allowing '12' itself but requiring '12='? And how about '<=12', '<12', '>12', '=12'?

kravciak commented:
@okurz thanks, I like your proposal '>=15', '=12-sp1'. I guess most used will be '>=',  '<' operators.
I will looks into implementation a bit later, thanks all for feedback!!

kravciak commented:
I implemented filter for version checks and removed sle_version_at_least from some modules.
It's before local runs, please check modified PR (mostly first commit).
For now there is only version check and I did not modify leap.
```
Format: [= > < >= <=] version [+]
Example: <=12-sp3 =12-sp1 <12 >=15
```

kravciak commented:
Please review

okurz commented on tests/installation/disable_grub_timeout.pm:
```diff
@@ -31,8 +31,9 @@ sub run {
     # Workaround for bsc#1070233: not update "Booting" option in upgrade mode
-    return record_soft_failure('bsc#1070233: Error if click on Booting option')
-      if (get_var('UPGRADE') && (!(is_sle && !sle_version_at_least('15')) || !(is_leap && !leap_version_at_least('15.0'))));
+    if (get_var('UPGRADE') && (!is_sle('<15') || !(is_leap && !leap_version_at_least('15.0')))) {
```
Now it should be possible to write `!is_sle('<15')` as `is_sle('15+')`, right?

kravciak replied:
`!is_sle('<15')` - is true for also for opensuse, so your 2 statements are not equal

kravciak commented:
Please ping me if you see any broken tests because of this PR - I am not watching sles functional group that much

rwx788 commented:
@kravciak I'll announce it to the team, so everyone is aware. Thanks!
